### PR TITLE
MSVC Compatibility

### PR DESCRIPTION
**Pull request by jseyster**

### CODE REVIEW

File: js/src/frontend/StencilXdr.h
Comment (BillyDonahue) [inline js/src/frontend/StencilXdr.h:30]: [optional] If there's an MSVC bug to track, that would be useful to cite here so we know when to remove this workaround.

File: js/src/vm/Interpreter.cpp
Comment (BillyDonahue) [inline js/src/vm/Interpreter.cpp]: It would be nice to label the else/endif so we don't get lost.

    #else  // have computed goto
    ...
    #endif // have computed goto

File: js/src/vm/Interpreter.cpp
Comment (BillyDonahue) [inline js/src/vm/Interpreter.cpp]: Many of these lines didn't need to change. Let's not insert a space after `#` in these directives to avoid noise.
If that's what they had upstream we can conform to it.

File: js/src/vm/Interpreter.cpp
Comment (BillyDonahue) [inline js/src/vm/Interpreter.cpp]: `PSEUDO` is unclear and specific to the only `OP` it happens to be used with. Maybe `CASE_NOT_COMPUTED` ?

File: js/src/vm/Interpreter.cpp
Comment (BillyDonahue) [inline js/src/vm/Interpreter.cpp]: This should not be indented.

File: js/src/wasm/WasmCode.h
Comment (BillyDonahue) [inline js/src/wasm/WasmCode.h]: I can't find documentation for the `_np` version of this intrinsic?

https://docs.microsoft.com/en-us/previous-versions/ttk2z1ws(v=vs.85)

Maybe we should call through the public `InterlockedCompareExchange64` instead, and let the compiler macros find the right C API "overload" for us?

File: js/src/wasm/WasmBaselineCompile.cpp
Comment (BillyDonahue) [inline js/src/wasm/WasmBaselineCompile.cpp]: remove blank.

Anonymous unions have certain special properties that could be erased by naming it. I think we can avoid doing so. A brace init of an anonymous union initializes its first member. So we can just swap the fields here:

```suggestion
    union {
      float f32;
      int32_t i32;
    } bits = {imm};
```

File: js/src/wasm/WasmBaselineCompile.cpp
Comment (BillyDonahue) [inline js/src/wasm/WasmBaselineCompile.cpp]: We can just swap the fields here.
```c++
    union {
      double f64;
      int64_t i64;
    } bits = {imm};
```

File: js/src/wasm/WasmFrame.h
Comment (BillyDonahue) [inline js/src/wasm/WasmFrame.h]: If we just put `allFlags` first in the union, we can initialize `_flags{expr}` with a `uint32_t expr`.

File: js/src/wasm/WasmFrame.h
Comment (BillyDonahue) [inline js/src/wasm/WasmFrame.h:320]: Alternatively I believe this could just change `{0}` to `{}` and rely on value-initialization of the struct?

File: js/src/frontend/TokenStream.cpp
Comment (BillyDonahue) [inline js/src/frontend/TokenStream.cpp]: [nit] Maybe put it on its own line so we diff more cleanly.

Maybe add a `static` to the definition of `ColumnChunkLength` instead? It shouldn't be necessary to add it to the capture list of this lambda.

File: js/src/jit/VMFunctionList-inl.h
Comment (BillyDonahue) [inline js/src/jit/VMFunctionList-inl.h:278]: Instead of modifying the string API for this special case, we can, I think, define ToStringSlowForVM here in this file instead?

File: js/src/vm/StringType.h
Comment (BillyDonahue) [inline js/src/vm/StringType.h]: That's not a concern local to this file. I think it belongs in the VMFunctionList-inl.h to which it's pertinent?

File: js/src/jit/CacheIRCompiler.cpp
Comment (BillyDonahue) [inline js/src/jit/CacheIRCompiler.cpp:6392]: same comment as 6382 should appear here.

**Review by BillyDonahue — CHANGES_REQUESTED** [review]

File: js/src/vm/Interpreter.cpp
Comment (jseyster) [reply js/src/vm/Interpreter.cpp]: I messed up here. I thought I had run this file through Firefox's `.clang-format`, but it turns out I didn't. After formatting, these lines actually have two spaces following the `#`. Personally, I think the extra readability from indenting nested directives and keeping formatting consistent is worth the diff noise, but I could remove the indentations if you think we should aggressively limit the diff size.

For reference, I copied this code from the older version of mozjs in our current master branch:
https://github.com/mongodb/mongo/blob/581c58c475a872e25b2e3bf7cf5ccd52425ef7c7/src/third_party/mozjs/extract/js/src/vm/Interpreter.cpp#L1755-L1789

File: js/src/vm/Interpreter.cpp
Comment (jseyster) [reply js/src/vm/Interpreter.cpp]: Good idea, thanks.

File: js/src/vm/Interpreter.cpp
Comment (jseyster) [reply js/src/vm/Interpreter.cpp]: Done. Thanks for the suggestion.

File: js/src/vm/Interpreter.cpp
Comment (jseyster) [reply js/src/vm/Interpreter.cpp]: Formatting this file with the mozjs `.clang-format` corrected this indentation for us. Nice catch, though!

File: js/src/wasm/WasmBaselineCompile.cpp
Comment (jseyster) [reply js/src/wasm/WasmBaselineCompile.cpp]: Very interesting! Thanks for catching this one and for the helpful workaround.

File: js/src/wasm/WasmBaselineCompile.cpp
Comment (jseyster) [reply js/src/wasm/WasmBaselineCompile.cpp]: Done.

File: js/src/wasm/WasmCode.h
Comment (jseyster) [reply js/src/wasm/WasmCode.h]: Thanks for this suggestion!

Good news:
I agree that this API function is more robust than relying on an intrinsic, and your suggestion to use `InterlockedCompareExchange64` led me to `InterlockedCompareExchangePointer`, which I think is even better for this purpose.
https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedcompareexchangepointer

Base news:
Both API functions require including `winnt.h`, which adds a bunch of `$define`s that conflict with identifiers throughout mozjs. I think the best way to deal with that would be to sequester `JumpTables::setJitEntryIfNull` to its own file. I've tried that out here. Let me know what you think.

File: js/src/wasm/WasmFrame.h
Comment (jseyster) [reply js/src/wasm/WasmFrame.h]: Done.

File: js/src/wasm/WasmFrame.h
Comment (jseyster) [reply js/src/wasm/WasmFrame.h:320]: Also a good idea. Between the two alternatives, I decided to go with reordering the Flags union so that we can initialize using a 0-value uint32_t. My reasoning is that it's conceivable that the compiler could init _just_ the bits that are specified in the struct to 0 and leave random junk in the other bits. Hard to imagine that happening in practice, but it's the best logic I could think of for choosing between the possible ways to initialize this variable.

File: js/src/frontend/StencilXdr.h
Comment (jseyster) [reply js/src/frontend/StencilXdr.h:30]: Agreed. I haven't filed any bugs yet, but I'd like to when I have a spare day to set up some minimal examples. Have you submitted MSVC bugs before? I'd be interested in the process. Thanks!

File: js/src/frontend/TokenStream.cpp
Comment (jseyster) [reply js/src/frontend/TokenStream.cpp]: Adding the `static` worked. Thanks!

File: js/src/jit/CacheIRCompiler.cpp
Comment (jseyster) [reply js/src/jit/CacheIRCompiler.cpp:6392]: Done.

File: js/src/jit/VMFunctionList-inl.h
Comment (jseyster) [reply js/src/jit/VMFunctionList-inl.h:278]: Done.

File: js/src/vm/StringType.h
Comment (jseyster) [reply js/src/vm/StringType.h]: Agreed.

**Review by jseyster — COMMENTED** [review]
Thanks for the fast review! I pushed two commits addressing your comments.

File: js/src/debugger/Object.cpp
Comment (mdashti) [inline js/src/debugger/Object.cpp:24]: Can you please add a comment as to why this header is needed? Similar to other imports in this file.

File: js/src/frontend/BytecodeEmitter.cpp
Comment (mdashti) [inline js/src/frontend/BytecodeEmitter.cpp:9776]: Interesting. Does MSVC get confused by this?

File: js/src/wasm/WasmBaselineCompile.cpp
Comment (BillyDonahue) [inline js/src/wasm/WasmBaselineCompile.cpp]:
```suggestion
    } bits{imm};
```
The note is incorrect as this isn't a default constructor.
Could fix it but I think maybe we don't need a note for the language feature?

File: js/src/vm/Interpreter.cpp
Comment (BillyDonahue) [inline js/src/vm/Interpreter.cpp]: These indents didn't need to change. I would like to minimize the diff if possible.

File: js/src/wasm/WasmCode-platform.cpp
Comment (BillyDonahue) [inline js/src/wasm/WasmCode-platform.cpp]: The using namespace directive can be dangerous.
I don't think this file needs it?

**Review by BillyDonahue — CHANGES_REQUESTED** [review]

File: js/src/frontend/StencilXdr.h
Comment (BillyDonahue) [reply js/src/frontend/StencilXdr.h:30]: Yes I've done it. You do it here. https://developercommunity.visualstudio.com/home

File: js/src/gc/RootMarking.cpp
Comment (mdashti) [inline js/src/gc/RootMarking.cpp:60]:
```suggestion
void trace(JSTracer*) { MOZ_ASSERT_UNREACHABLE(); }
```

I'm pretty sure that `= delete` also leads to a compilation issue on GCC. Isn't it better to just use a default unreachable implementation?

**Review by mdashti — APPROVED** [review]
Thanks @jseyster. Beside some minor comments (and the comments already added by Billy), looks good to me.

File: js/src/gc/RootMarking.cpp
Comment (BillyDonahue) [reply js/src/gc/RootMarking.cpp:60]: No it isn't better *IF* it can be avoided. A use of an `=delete` function cannot be compiled.
Maybe I don't know exactly what you mean by a "default unreachable implementation". This macro is a runtime failure, so that's not as good as a compile failure. But if there is actually a use of the function in some way that is implementation-dependent, then I agree we shouldn't leave it to chance.

I'm curious about how this ended up being a compiler-dependent issue, but not enough to block the PR.

File: js/src/debugger/Object.cpp
Comment (BillyDonahue) [reply js/src/debugger/Object.cpp:24]: I kind of dislike those comments and Google style discourages them because they are almost immediately out of date.

**Comment by BillyDonahue** [comment]
As far as clant-format and indentation. Yes. I think especially for vendored products like this, we should be VERY aggressively limiting diff size so that we can upstream our patches if needed, and to facilitate cleaner future merges as upstream evolves.

I don't personally like indenting after the `#`. It can make preprocessor directives interleave with non-directive lines of code, and it means that adding an outer #if/#endif will cause interior directives to shift, which causes noise later.

File: js/src/debugger/Object.cpp
Comment (jseyster) [reply js/src/debugger/Object.cpp:24]: Good eye catching that this header should have a comment to match the style of the others. I tend to agree with Billy, though, that the comment would not be so helpful here. The classes that we need from Environment.h are not directly used in this file, anyway.

File: js/src/frontend/BytecodeEmitter.cpp
Comment (jseyster) [reply js/src/frontend/BytecodeEmitter.cpp:9776]: Yes. I might have forgotten to mention that the git commit comments include details about some of the changes. From the comment for this change:

"Simplify calls to mozjs StringBuffer::append(const char (&array)[ArrayLength])

"In Clang/g++, type deduction figures out that both sides of the ternary operator (see the diff) can be a char[8] array, which allows overload resolution to choose the StringBuffer::append(const char (&array)[ArrayLength]) overload, where ArrayLength is a template parameter with value 8.

"MSVC's type deduction cannot make this inference, however, so we need to interchange the outer StringBuffer::append() call with the inner ternary operator, simplifying inference. The resulting code is arguably more robust: before this change, the code was only legal because the ".getter" and ".setter" strings happen to have the same length."

File: js/src/frontend/StencilXdr.h
Comment (jseyster) [reply js/src/frontend/StencilXdr.h:30]: Thanks for the info! I will try to get feedback submitted after these changes are ready.

File: js/src/gc/RootMarking.cpp
Comment (jseyster) [reply js/src/gc/RootMarking.cpp:60]: Agreed that this isn't ideal. The advantage of using the #ifdef is that our non-Windows builders continue to statically validate that there is no attempt to call this deleted function, which gives us confidence that we'll never reach this MOZ_ASSERT_UNREACHABLE() on the Windows build.

I haven't encountered any issue with this deleted function in gcc: it compiles correctly in Evergreen for me.

Unfortunately, I wasn't able to narrow this down to a simple enough case that I could definitively pinpoint the source of the problem. MSVC is more conservative about which call sites it think may be calling this function. From my inspection, I feel confident that this function cannot be called and that the compiler error is a false alarm, but I will have to go back and run the compiler again to get the exact message. (That will take me an hour or so to do, because my Windows Evergreen instance somehow became inaccessible :( ).

File: js/src/vm/Interpreter.cpp
Comment (jseyster) [reply js/src/vm/Interpreter.cpp]: Done.

File: js/src/wasm/WasmBaselineCompile.cpp
Comment (jseyster) [reply js/src/wasm/WasmBaselineCompile.cpp]: I deleted the comment.

File: js/src/wasm/WasmCode-platform.cpp
Comment (jseyster) [reply js/src/wasm/WasmCode-platform.cpp]: We need access to the {{js::wasm}} namespace to resolve {{JumpTable}} in the definition of {{JumpTables::setJitEntryIfNull}} below.

Instead of a {{using}} directive, I updated this code to explicitly place the function definition in the namespace, which I think will be safer.

**Review by jseyster — COMMENTED** [review]
I pushed a new commit responding to the remaining comments.

Note that Mohammad pushed a merge commit to this branch, so some unrelated changes will appear in the "Files changed" section. I recommend looking specifically at the new commit to see the recent changes to the branch. Thanks!

File: mfbt/EndianUtils.h
Comment (jseyster) [inline mfbt/EndianUtils.h:104]: I don't have a definitive source for this, but it should at least apply to platforms that MongoDB supports.

Additionally, the "possible implementation" of {{std::endian}} strongly implies that Windows targets expect a little-endian architecture.
https://en.cppreference.com/w/cpp/types/endian

Once we are fully transitioned to C++20, we'll have the static_assert to keep this statement honest.

**Review by jseyster — COMMENTED** [review]
I modified EndianUtils.h to detect MSVC in order to address this comment:
https://github.com/10gen/mongo/pull/2878#discussion_r807333109

The proposed implementation will work in C++17 with additional checking enabled in C++20.

File: js/src/wasm/moz.build
Comment (jseyster) [inline js/src/wasm/moz.build:52]: In the previous commit where I spun `WasmCode-platform.cpp` into a new file, I forgot to add it to the build system, which broke compilation outside of the `mongo` code base.

I'm adding it here instead of directly to `UNIFIED_SOURCES`, because adding it to `UNIFIED_SOURCES` would require us to update the `src/third_party/mozjs/platform/<ARCH>/<PLATFORM>/build/Unified_cpp_js_src_wasm*.cpp` files for every platform.

@mdashti Unless we decide to bite the bullet and add this to the `UNIFIED_SOURCES` list, we'll have to also add this file to `src/third_party/mozjs/SConstruct`.

**Review by jseyster — COMMENTED** [review]
Posting two more patches to address issues I hit while attempting a standalone build of Spidermonkey.

File: js/src/wasm/moz.build
Comment (mdashti) [reply js/src/wasm/moz.build:52]: @jseyster IMO, it's OK to keep it outside `UNIFIED_SOURCES`, as it requires a lot of effort to regenerate the configurations for all platform.

File: mfbt/EndianUtils.h
Comment (mdashti) [reply mfbt/EndianUtils.h:104]: @jseyster This change looks good to me.

**Comment by jseyster** [comment]
Thanks for all the helpful feedback on this change! I rebased this branch, cleaned up the commits, and pushed everything to the main branch: `spidermonkey-esr91.3-cpp-only`. I'm closing this out as completed.

**Pull request closed without merging**
